---
title: "assignment"
author: "José F Carreño"
date: '2022-04-20'
output: html_document
---

# Load the necessary libraries

```{r}
suppressPackageStartupMessages({
  library(epiwraps)
  library(AnnotationHub)
  library(MotifDb)
  library(memes)
  library(universalmotif)
  library(ensembldb)
  library(ggplot2)
})
ah <- AnnotationHub(localHub=TRUE)
genome <- ah[["AH88477"]]
```

# Necessary functions for further analysis: findMotifInstances

```{r}
#' findMotifInstances
#'
#' A wrapper around TFBSTools for scanning motif occurence, handling the 
#' coordinate conversion as `memes` does.
#' 
#' @param seqs A set of sequences, e.g. `DNAStringSet`, optionally with 
#'   coordinate as names to enable conversion.
#' @param motif A motif, in any format recognized by `universalmotif`
#' @param keepMatchedSeq Logical; whether to keep the matched sequence.
#' @param ... Passed to `TFBSTools::searchSeq`; can for instance be used to set
#'   the number of threads to use, e.g. with `mc.cores=2`
#'
#' @return A `GRanges` object
findMotifInstances <- function(seqs, motif, keepMatchedSeq=FALSE, ...){
  motif <- universalmotif::convert_motifs(motif, "TFBSTools-PWMatrix")
  x <- suppressWarnings(TFBSTools::searchSeq(motif, subject=seqs))
  peaks <- strsplit(gsub("-",":",names(seqs)), ":")
  if(all(lengths(peaks)==3)){ # convert relative coordinates to absolute
    chrs <- sapply(peaks,FUN=function(x) x[1])
    offsets <- sapply(peaks,FUN=function(x) as.integer(x[2]))-1L
    i <- rep(seq_along(x),lengths(x))
    return(GRanges(chrs[i], 
                   IRanges(offsets[i]+as.integer(unlist(lapply(x, start))),
                           offsets[i]+as.integer(unlist(lapply(x, end)))),
                   strand=unlist(lapply(x, strand)),
                   score=as.numeric(unlist(lapply(x, FUN=function(x) score(x))))
                   ))
  }
  x <- as(x, "GRanges")
  score(x) <- x$absScore
  keepFields <- c("score","relScore")
  if(keepMatchedSeq) keepFields <- c(keepFields, "siteSeqs")
  mcols(x) <- mcols(x)[,intersect(colnames(mcols(x)), keepFields)]
  x
}
```


# Download the necessary files

NOTE: We are asked for the insertion sites (i.e. cuts) of each factor. Hence we do not need the coverage information. We will include the signal of the full cuts for comparison of the individual signals against the complete signal.

```{r, eval=FALSE}
download.file("https://ethz-ins.org/content/w7/atac.chr19.bam", "atac.chr19.bam", mode="wb")
download.file("https://ethz-ins.org/content/w7/atac.chr19.bam.bai", "atac.chr19.bam.bai", mode="wb")
# if you're on a slow computer, you might also want to download the tracks that we're generating below:
tracks <- c("full_cuts.bw", "NF_cuts.bw", "mono_cuts.bw")
for(f in tracks){
  download.file(paste0("https://ethz-ins.org/content/w7/",f), f, mode="wb")
}
```

## Process the downloaded files. Obtaining the traks

This chunk is not run as the fragments have been previously downloaded.

```{r, eval=FALSE}
bam <- "atac.chr19.bam"
# create a track using all fragments, computing the coverage at each position (and force Ensembl naming)
bam2bw(bam, output_bw = "full_cov.bw", paired=TRUE, binWidth=1L, forceSeqlevelsStyle = "Ensembl")
# create a track using all fragments, but computing the number of cuts/insertion sites at each position
bam2bw(bam, output_bw = "full_cuts.bw", paired=TRUE, binWidth=1L, type="ends", forceSeqlevelsStyle = "Ensembl")
# create a track using only nucleosome-free fragments, computing coverage
bam2bw(bam, output_bw = "NF_cov.bw", paired=TRUE, binWidth=1L, minFragLength=30, 
       maxFragLength=115, forceSeqlevelsStyle = "Ensembl")
# create a track using only nucleosome-free fragments, computing cuts
bam2bw(bam, output_bw = "NF_cuts.bw", paired=TRUE, binWidth=1L, minFragLength=30,
       maxFragLength=115, type="ends", forceSeqlevelsStyle = "Ensembl")
# create a track using only mono-nucleosome fragments, computing coverage
bam2bw(bam, output_bw = "mono_cov.bw", paired=TRUE, binWidth=1L, minFragLength=130,
       maxFragLength=220, forceSeqlevelsStyle = "Ensembl")
# create a track using only mono-nucleosome fragments, computing cuts
bam2bw(bam, output_bw = "mono_cuts.bw", paired=TRUE, binWidth=1L, minFragLength=130,
       maxFragLength=220, type="ends", forceSeqlevelsStyle = "Ensembl")
```


# Find the motifs of interest

Finding the motif of the REST factor:

```{r}
# we search for "REST" in the motif database
motifs <- query(MotifDb, "REST")
# there are several matching motifs:
names(motifs)
# we select one:
motif_REST <- motifs[["Mmusculus-HOCOMOCOv10-REST_MOUSE.H10MO.A"]]
# we visualize it:
view_motifs(motif_REST)
```

Finding the motif of the GCR factor:

```{r}
# we search for "GCR" in the motif database
motifs <- query(MotifDb, "GCR")
# there are several matching motifs:
names(motifs)
# we select one:
motif_GCR <- motifs[["Mmusculus-HOCOMOCOv10-GCR_MOUSE.H10MO.C"]]
# we visualize it:
view_motifs(motif_GCR)
```

Obtaining the sequence of the chromosome 19:

```{r}
chr19 <- GRanges("19", IRanges(1L,seqlengths(genome)["19"]))
chr19 <- setNames(getSeq(genome, chr19), "19")
```

Finding the motifs instances through the genome and keep only those with high confidence:

```{r}
moi_REST <- findMotifInstances(chr19, motif_REST, mc.cores=2) # running with 2 threads
moi_GCR <- findMotifInstances(chr19, motif_GCR, mc.cores=2) # running with 2 threads

moi_REST_conf <- moi_REST[moi_REST$score>15]
moi_GCR_conf <- moi_GCR[moi_GCR$score>15]
```

# Plot the signal tracks around the motifs of interest

We are asked for the nucleosome-free and nucleosome-containing fragments.

```{r}
# we prepare the list of tracks
tracks <- c("Full cuts"="full_cuts.bw", "NF cuts"="NF_cuts.bw", "Mono cuts"="mono_cuts.bw")

# extract signals around the motif occurences
# we zoom in to 500bp around the motif centers, in windows of 5bp
sm_REST <- signal2Matrix(tracks, moi_REST_conf, w=5, extend=500)
sm_GCR <- signal2Matrix(tracks, moi_GCR_conf, w=5, extend=500)

# plot the signals:
# plotEnrichedHeatmaps(sm_REST, trim=0.95)
# plotEnrichedHeatmaps(sm_GCR, trim=0.95)
```

Histogram of the REST factor around the nucleosome free and nucleosome containing fragments:

```{r}
# we renormalize using border background
smb_REST <- renormalizeBorders(sm_REST)
plotEnrichedHeatmaps(smb_REST, trim=0.95)
```

Histogram of the GCR factor around the nucleosome free and nucleosome containing fragments:

```{r}
smb_GCR <- renormalizeBorders(sm_GCR)
plotEnrichedHeatmaps(smb_GCR, trim=0.95)
```
Comment: After establishing different thresholds for the score of the found motifs, 15 has been selected. Nevertheless, the signal obtained for the GCR is still bad quality. 