---
title: "assignment"
author: "José F Carreño"
date: '2022-04-03'
output: html_document
---

```{r}
suppressPackageStartupMessages({
  library(GenomicRanges)
  library(memes)
  library(Biostrings)
  library(AnnotationHub)
  library(MotifDb)
  library(universalmotif)
  library(PWMEnrich)
  library(ggplot2)
})

ah <- AnnotationHub(localHub=TRUE)
#genome <- ah[["AH89211"]]
genome <- ah[["AH88477"]] #NOTE: The genome must be in FASTA format to be feeded to the memes library
```

# Select a transcription factor

I have selected the GATA1 transcription factor.The cell line is G1E-ER4. Download the peaks for this factor: 

```{r}
download.file("https://www.encodeproject.org/files/ENCFF045RXZ/@@download/ENCFF045RXZ.bed.gz", "ENCFF045RXZ.bed.gz")
peaks <- rtracklayer::import("ENCFF045RXZ.bed.gz", format="NarrowPeak")
seqlevelsStyle(peaks) <- "Ensembl"
```

```{r}
peak_centers <- resize(peaks, fix="center", width=100)
# we get the sequences corresponding to the peak centers:
peak_seqs <- memes::get_sequence(peak_centers, genome)
peak_seqs
```


# Motif scanning

To obtain the motif of interest: 

```{r}
# we search for "GATA1" in the motif database
motifs <- query(MotifDb, "GATA1")
names(motifs)
```

```{r}
# We must select one of the given motifs
motif <- motifs[["Mmusculus-jaspar2018-Tal1::Gata1-MA0140.1"]]
# we visualize it:
view_motifs(motif)
```


To scan the motif along the genome we use the function given in class

```{r}
#' findMotifInstances
#'
#' A wrapper around TFBSTools for scanning motif occurence, handling the 
#' coordinate conversion as `memes` does.
#' 
#' @param seqs A set of sequences, e.g. `DNAStringSet`, optionally with 
#'   coordinate as names to enable conversion.
#' @param motif A motif, in any format recognized by `universalmotif`
#' @param keepMatchedSeq Logical; whether to keep the matched sequence.
#' @param ... Passed to `TFBSTools::searchSeq`; can for instance be used to set
#'   the number of threads to use, e.g. with `mc.cores=2`
#'
#' @return A `GRanges` object
findMotifInstances <- function(seqs, motif, keepMatchedSeq=FALSE, ...){
  motif <- universalmotif::convert_motifs(motif, "TFBSTools-PWMatrix")
  x <- suppressWarnings(TFBSTools::searchSeq(motif, subject=seqs))
  peaks <- strsplit(gsub("-",":",names(seqs)), ":")
  if(all(lengths(peaks)==3)){ # convert relative coordinates to absolute
    chrs <- sapply(peaks,FUN=function(x) x[1])
    offsets <- sapply(peaks,FUN=function(x) as.integer(x[2]))-1L
    i <- rep(seq_along(x),lengths(x))
    return(GRanges(chrs[i], 
                   IRanges(offsets[i]+as.integer(unlist(lapply(x, start))),
                           offsets[i]+as.integer(unlist(lapply(x, end)))),
                   strand=unlist(lapply(x, strand)),
                   score=as.numeric(unlist(lapply(x, FUN=function(x) score(x))))
                   ))
  }
  x <- as(x, "GRanges")
  score(x) <- x$absScore
  keepFields <- c("score","relScore")
  if(keepMatchedSeq) keepFields <- c(keepFields, "siteSeqs")
  mcols(x) <- mcols(x)[,intersect(colnames(mcols(x)), keepFields)]
  x
}
```


```{r}
motif_instances <- findMotifInstances(peak_seqs, motif, mc.cores=2)
```

```{r}
mmusculus <- import(genome, "2bit", which = as(seqinfo(genome), "GenomicRanges"))
motif_instances_genome <- findMotifInstances(mmusculus, motif, mc.cores=2)
```

```{r}
motif_with_peaks = overlapsAny(motif_instances_genome, peaks)
```

# Answer the specific questions

Q1: Of the `r length(peaks)` peaks, `r length(motif_instances)` (`r length(motif_instances)/length(peaks)*100` %) contain a motif.

Q2: Of the `r length(motif_instances_genome)` motif instances in the genome, `r sum(motif_with_peaks)` (`r (sum(motif_with_peaks)/length(motif_instances_genome))*100` %) have a peak
